hermes/probe/checknil/check_nil.go

```diff
@@ -0,0 +1,332 @@
+//	- bool: returns true if the hash matches, else false.
+func (n *nilFile) validateChecksum() bool {
+	if n.hash == "" || n.name == "" {
+		return false
+	}
+
+	// Here the checksum in the filename is being validated against the file contents
+	// This will return false if the file contents have been updated
+	// without the file name being changed accordingly.
+	return n.hash == n.name[len("Hermes_NIL_"):]
+}
+
+// reader returns the file contents as a reader.
+// Returns:
+// - reader: returns the file contents as a reader.
+func (n *nilFile) reader() io.Reader {
+	return bytes.NewReader(n.contents)
+}
+
+// CheckNilFile checks the existence and validity of the Nil file on the target
+// storage system. It also reads the intent from the Nil file to identify if
+// Hermes failed to complete the last operation, and completes it if necessary.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+func CheckNilFile(ctx context.Context, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	bucket := target.Target.GetBucketName()
+
+	file, err := checkNilFileExists(ctx, bucket, target, client, logger)
+	if err.Status != metrics.Success {
+		return err
+	}
+
+	if file, err = readNilFile(ctx, file, bucket, target, client, logger); err.Status != metrics.Success {
+		return err
+	}
+
+	if err = verifyNilFile(ctx, file, target, client, logger); err.Status != metrics.Success {
+		return err
+	}
+
+	logger.Infof("CheckNilFile() completed successfully, nil file is valid.")
+	return probe.NewProbeError(metrics.Success, nil)
+}
+
+// checkNilFileExists checks that the nil file exists on the target system.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- nilFile: a nilFile struct with the file name recorded
+//	- err: error exit statuses are well defined in metrics.go
+func checkNilFileExists(ctx context.Context, bucket string, target *probe.Target, client stiface.Client, logger *logger.Logger) (*nilFile, *probe.ProbeError) {
+	file := &nilFile{}
+
+	err := RecordAPILatency(target.LatencyMetrics.APICallLatency, metrics.APIListFiles, func() *probe.ProbeError {
+		objIter := client.Bucket(bucket).Objects(ctx, &storage.Query{Prefix: "Hermes_Nil"})
+		for {
+			// TODO(evanSpendlove): Review how Nil file is updated and consider multiple nil files being present.
+			// If one, success
+			// If two -> verify whih one reflects the current state of the system and delete the other.
+			// If two and only differ by intent, prefer the one with intent.
```

Comment:
This is especially nitpicky, but you probably want to check whether the intent makes sense when comparing these.

Also, you could add a datetime to the nil file to help with this?

Reply:
I've added a TODO for this.